### ISSUE TITLE
Enhanced LIMIT usage

### ISSUE THREAD

**Issue by ramazanpolat** [ClickHouse/ClickHouse#40965]
When writing complex queries that involves multiple JOINs we usually want to work on a subset of tables.

```sql
SELECT *
FROM default.intraday_metrics
LIMIT 10

/*
┌─asset_id─┬─metric_id─┬──────────────────dt─┬─value─┬─────────computed_at─┐
│     2503 │       250 │ 2008-12-10 00:00:00 │     0 │ 2020-02-28 15:18:42 │
│     2503 │       250 │ 2008-12-10 00:05:00 │     0 │ 2020-02-28 15:18:42 │
│     2503 │       250 │ 2008-12-10 00:10:00 │     0 │ 2020-02-28 15:18:42 │
│     2503 │       250 │ 2008-12-10 00:15:00 │     0 │ 2020-02-28 15:18:42 │
│     2503 │       250 │ 2008-12-10 00:20:00 │     0 │ 2020-02-28 15:18:42 │
│     2503 │       250 │ 2008-12-10 00:25:00 │     0 │ 2020-02-28 15:18:42 │
│     2503 │       250 │ 2008-12-10 00:30:00 │     0 │ 2020-02-28 15:18:42 │
│     2503 │       250 │ 2008-12-10 00:35:00 │     0 │ 2020-02-28 15:18:42 │
│     2503 │       250 │ 2008-12-10 00:40:00 │     0 │ 2020-02-28 15:18:42 │
│     2503 │       250 │ 2008-12-10 00:45:00 │     0 │ 2020-02-28 15:18:42 │
└──────────┴───────────┴─────────────────────┴───────┴─────────────────────┘
*/
```
Then we iteratively add JOINs:
```sql
SELECT * 
FROM default.intraday_metrics AS im
LIMIT 10
INNER JOIN default.asset_metadata AS am ON im.asset_id = am.asset_id
INNER JOIN default.metric_metadata AS mm ON im.metric_id = mm.metric_id
```
But this query fails because of the LIMIT 10 line.

In order to see result, we have to remove the `LIMIT`, which in turn make CH to scan the whole table and start joining the tables. Obviously this takes too much time and memory. But we just need to see a sample output, 10 rows is enough for us to see if we wrote the query correctly or even see the data to decide if everything is OK. Therefore we rewrite the query like this:

```sql
SELECT * 
FROM (SELECT * FROM default.intraday_metrics LIMIT 10) im
INNER JOIN (SELECT * FROM default.asset_metadata LIMIT 100) AS am ON im.asset_id = am.asset_id
INNER JOIN (SELECT * FROM default.metric_metadata LIMIT 100) AS mm ON im.metric_id = mm.metric_id
```

As you've guessed, this makes the query harder to write and read. Not only that, if everything works fine, then we have to replace all `… (SELECT … LIMIT ?)` to actual tables.

Another example:
```sql
SELECT * 
FROM default.intraday_metrics
LIMIT 100
ORDER BY asset_id
```

This query also fails. Because LIMIT is used only after all query is run and the resultset is generated. But here, the intention is to just take the first 100 rows of the table and THEN order by asset_id, which is pretty straightforward and intuitive. Unfortunately standard SQL does not conform to this.

I believe that’s why usage of the `system.numbers` evolved to `SELECT * FROM numbers(10)` because of this reason and shows that there is a need for this feature.

**Describe the solution you'd like**
Being able to do it for any table like this would be nice:
```sql
SELECT *
FROM my_table(100)
```

**Describe alternatives you've considered**
Or more explicitly:
```sql
SELECT *
FROM my_table(limit=100)
```

**Comment by ramazanpolat** [comment]
Related to #40907

**Comment by UnamedRus** [comment]
As another idea:

There is SAMPLE, which can limit amount of rows in sample to read.
And it's being defined just after table name (just like LIMIT proposition here)

We may need to make SAMPLE work even without explicitly defined SAMPLE KEY and/or ClickHouse trying to take correctly distributed sample .

**Comment by ramazanpolat** [comment]
@UnamedRus Although using `SAMPLE` would also be a great option, I still think that doing it with `LIMIT` will be more intuitive. But again, I'd also accept `SAMPLE` as a viable solution.

**Comment by alexey-milovidov** [comment]
The variant with parenthesis like `FROM my_table(100)` is too ambiguous, it can be confused with table functions, and it gives the one use case (taking a few records) too much favor.

The variant with an additional LIMIT section like `SELECT FROM table *LIMIT* 10 ORDER BY x LIMIT 10` may be ok, but looks too surprising.

There is a variant with another additional clause with a different name like `SELECT FROM table *TAKE* 10 ORDER BY x LIMIT 10`.

There is also a variant with repurposing the SAMPLE section for this like `SELECT FROM table SAMPLE *ANY* 10 ORDER BY x LIMIT 10`.

I would postpone implementation before more motivation for this task will appear.

**Comment by UnamedRus** [comment]
BTW, now it can be done via parametrized view

```sql
CREATE VIEW numbers_pv
AS SELECT *
FROM system.numbers
LIMIT {limit:UInt32}

SELECT * FROM numbers_pv(limit=100);

     ┌─number─┐
  1. │      0 │
  2. │      1 │
  3. │      2 │
  4. │      3 │
  5. │      4 │
  6. │      5 │
  7. │      6 │
 
...
 94. │     93 │
 95. │     94 │
 96. │     95 │
 97. │     96 │
 98. │     97 │
 99. │     98 │
100. │     99 │
     └─number─┘
```
